### PR TITLE
test: add method to create the sui coin metadata

### PR DESCRIPTION
**Pull request by josemvcerqueira**
## Description 

It adds a test only function to create Sui currency in tests.

## Test Plan 

How did you test the new or updated feature?

We added a a file to test the function.

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

**Comment by amnn** [comment]
Hi @josemvcerqueira, have you seen the `coin::mint_for_testing` and `balance::create_for_testing` helpers? They allow you to create coins and balances for any coin type in the test environment:

https://github.com/MystenLabs/sui/blob/473e4e25615fb47cd788aaba849a372189df52bc/crates/sui-framework/packages/sui-framework/sources/coin.move#L409-L413

https://github.com/MystenLabs/sui/blob/473e4e25615fb47cd788aaba849a372189df52bc/crates/sui-framework/packages/sui-framework/sources/balance.move#L118-L122

**Comment by josemvcerqueira** [comment]
> Hi @josemvcerqueira, have you seen the `coin::mint_for_testing` and `balance::create_for_testing` helpers? They allow you to create coins and balances for any coin type in the test environment:
> 
> https://github.com/MystenLabs/sui/blob/473e4e25615fb47cd788aaba849a372189df52bc/crates/sui-framework/packages/sui-framework/sources/coin.move#L409-L413
> 
> https://github.com/MystenLabs/sui/blob/473e4e25615fb47cd788aaba849a372189df52bc/crates/sui-framework/packages/sui-framework/sources/balance.move#L118-L122

Hello @amnn, yes I have.

These updates are to allow developers to access the CoinMetadata in tests.

Thank you for the response.

File: crates/sui-framework/packages/sui-framework/sources/sui.move [hunk 1]
```diff
@@ -54,4 +54,9 @@ module sui::sui {
     public entry fun transfer(c: coin::Coin<SUI>, recipient: address) {
         transfer::public_transfer(c, recipient)
     }
+
+    #[test_only]
+    public fun new_for_testing(ctx: &mut TxContext): Balance<SUI> {
```
Comment (amnn) [inline crates/sui-framework/packages/sui-framework/sources/sui.move:59]: Could you add a doc-string here explaining the motivation for this function vs using the coin and balance test functions?

**Review by amnn — COMMENTED** [review]
Got it, thanks for clarifying @josemvcerqueira ! This seems reasonable to me, but let me get someone with some more Move expertise to double check.

**Comment by amnn** [comment]
@damirka / @tzakian / @tnowacki, could you double check for me that this seems okay? It looks fine to me, and is serving a niche that I can't think of another way of serving, but given that it's related to `SUI` initialisation, wanted to get some more eyes on it.

File: crates/sui-framework/packages/sui-framework/sources/sui.move (hunk 1 above)
Comment (josemvcerqueira) [reply crates/sui-framework/packages/sui-framework/sources/sui.move:64]: I added a doc-string explaining the reason behind this update. Please let me know if it is clear.

Comment (amnn) [reply crates/sui-framework/packages/sui-framework/sources/sui.move:64]: Looks great, thanks!

**Review by damirka — COMMENTED** [review]
Thanks for submitting the PR @josemvcerqueira! 

The method you're suggesting won't allow simple tests without TestScenario, and relies on storage. Do you think it would work better if `sui::coin` had a method like `new_coin_metadata_for_testing`, and `sui::sui` could implement it in `test_only` mode? Eg. `sui::sui::new_metadata_for_testing()`?

**Comment by josemvcerqueira** [comment]
> Thanks for submitting the PR @josemvcerqueira!
> 
> The method you're suggesting won't allow simple tests without TestScenario, and relies on storage. Do you think it would work better if `sui::coin` had a method like `new_coin_metadata_for_testing`, and `sui::sui` could implement it in `test_only` mode? Eg. `sui::sui::new_metadata_for_testing()`?

Hello Damir, thank you for your suggestion.

I believe this would be better yes. Should the new `sui::sui::new_metadata_for_testing()`be a friend-only function for `sui::sui` similar to `sui::balance::destroy_supply`?

File: crates/sui-framework/packages/sui-framework/tests/sui_tests.move [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      let sui_coin_metadata = test_scenario::take_immutable<coin::CoinMetadata<sui::SUI>>(scenario_mut);
+
+      assert_eq(coin::get_decimals(&sui_coin_metadata), SUI_DECIMALS);
+
+      test_scenario::return_immutable(sui_coin_metadata);
+    };
+
+    test_scenario::end(scenario);
+  }
+}
```
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/tests/sui_tests.move:35]: nit, newline

File: crates/sui-framework/packages/sui-framework/tests/sui_tests.move [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      let sui_balance = sui::new_for_testing(test_scenario::ctx(scenario_mut));
+      assert_eq(balance::destroy_for_testing(sui_balance), SUI_TOTAL_SUPPLY);
+    };
+
+    test_scenario::next_tx(scenario_mut, SENDER);
+    {
+      let sui_coin_metadata = test_scenario::take_immutable<coin::CoinMetadata<sui::SUI>>(scenario_mut);
+
+      assert_eq(coin::get_decimals(&sui_coin_metadata), SUI_DECIMALS);
+
```
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/tests/sui_tests.move:29]:
```suggestion
      assert_eq(coin::get_decimals(&sui_coin_metadata), SUI_DECIMALS);
```

**Comment by josemvcerqueira** [comment]
I updated the PR based on @damirka's feedback. Please let me know if there are additional changes :)

**Pull request closed without merging**